// @grammyjs/nestjs/src/nestjs-grammy.module.js
"use strict";
var NestjsGrammyModule_1;
Object.defineProperty(exports, "__esModule", { value: true });
exports.NestjsGrammyModule = void 0;
const tslib_1 = require("tslib");
const common_1 = require("@nestjs/common");
const nestjs_grammy_core_module_1 = require("./nestjs-grammy-core.module");
let NestjsGrammyModule = NestjsGrammyModule_1 = class NestjsGrammyModule {
    static forRoot(options) {
        return {
            module: NestjsGrammyModule_1,
            imports: [nestjs_grammy_core_module_1.GrammyCoreModule.forRoot(options)],
            exports: [nestjs_grammy_core_module_1.GrammyCoreModule],
        };
    }
    static forRootAsync(options) {
        return {
            module: NestjsGrammyModule_1,
            imports: [nestjs_grammy_core_module_1.GrammyCoreModule.forRootAsync(options)],
            exports: [nestjs_grammy_core_module_1.GrammyCoreModule],
        };
    }
};
NestjsGrammyModule = NestjsGrammyModule_1 = tslib_1.__decorate([
    (0, common_1.Module)({
        controllers: [],
        providers: [],
        exports: [],
    })
], NestjsGrammyModule);
exports.NestjsGrammyModule = NestjsGrammyModule;
